README.md
# calrom

![Build Status](https://github.com/calendarium-romanum/calrom/actions/workflows/ci.yml/badge.svg)
[![Gem Version](https://badge.fury.io/rb/calrom.svg)](https://badge.fury.io/rb/calrom)

Command line utility providing access to the Roman Catholic
liturgical calendar (post-Vatican II).

Built on top of the [calendarium-romanum][caro] Ruby gem.

## Installation

calrom is a Ruby gem:

`$ gem install calrom`

## Usage

### Specifying date range

Print liturgical calendar for the current month (default):

`$ calrom`

... for a specified month of the current year:

`$ calrom -m 9`

... for a specified month of another year:

`$ calrom -m 1 2028` or `$ calrom 1 2028`

... for a whole year:

`$ calrom 2017`

... for the current year:

`$ calrom -y`

... for a specified date:

`$ calrom 2028-01-15`

... for an arbitrary date range:

`$ calrom 2028-01-15 2028-03-07`

### Day and celebration filtering

In addition to specifying date range, filtering is a way to further refine
the selection of data to be displayed. Options `--day-filter=` and `--celebration-filter=`
both accept a snippet of Ruby code, which is then `eval`ed in context of each
`CalendariumRomanum::Day` / `CalendariumRomanum::Celebration` within the selected date range
and only days/celebrations for which the expression evaluates truthy are displayed.

Display only Saturdays:

`$ calrom --day-filter='date.saturday?'`

Display all Mondays with a celebration of a rank higher than memorial:

`$ calrom --day-filter='date.monday?' --day-filter='celebrations[0].rank > MEMORIAL_GENERAL'`

(As you can see, `calendarium-romanum` constants like ranks or seasons are available
as top-level constants. Noone likes extensive writing in the terminal.)

Display only ferials:

`$ calrom --celebration-filter='ferial?'`

Display only celebrations in green:

`$ calrom --celebration-filter='colour == GREEN'`

The options can be combined and used repeatedly to narrow the selection down as needed.

### Selecting calendar

There are a few calendars bundled in calrom (actually in the calendarium-romanum gem)
and ready to use. List them:

`$ calrom --calendars`

Each entry of the listing contains an ID of the calendar, it's name and language code.
Use calendar ID to request General Roman Calendar in Latin:

`$ calrom --calendar=universal-la`

You can prepare [your own calendar data][carodata] and load them:

`$ calrom --calendar=path/to/my_calendar.txt`

If you specify more than one calendar, they are loaded "layered" one over another
(from left to right), which comes in handy when extending a general calendar
with just a few additional and/or differing celebrations, e.g. solemnities (titular, dedication)
of the local church:

`$ calrom --calendar=universal-la --calendar=path/to/our_local_celebrations.txt`

Please note that specifying more than one calendar disables automatic loading of
parent calendars. If any of the listed calendars extends a parent calendar,
the parent either has to be explicitly listed using the `--calendar` option in order to be loaded,
or automatic parent loading has to be explicitly enabled using the `--load-parents` option.

Limited support for remote calendars is provided. Calendar URL from the
[Liturgical Calendar API](http://calapi.inadiutorium.cz/) or a compatible calendar API
is accepted as a value of `--calendar=`:

`$ calrom --calendar=http://calapi.inadiutorium.cz/api/v0/en/calendars/general-la`

### Data presentation settings

Print detailed listing:

`$ calrom -l`

Print current day in a condensed format (intended mainly for use cases like
window manager toolbars):

`$ calrom --format=condensed --today`

Disable colours:

`$ calrom --no-color`

Machine-readable output formats:

`$ calrom --format=json` - prints JSON array containing one object per day.
The object contents mimick output of the [Church Calendar API v0][calapidoc].

`$ calrom --format=csv` - prints a CSV, one celebration per line
(i.e. there is one or more lines for each liturgical day).

### Configuration files

`calrom` looks for configuration files `/etc/calromrc` and `~/.calromrc`.
They are processed in this order and both are used if available.
Their syntax is that of shell options and arguments (with the sole exception that newline
is not considered end of shell input, but generic whitespace), supported are all options and arguments
accepted by the command.

If a custom configuration file location is specified on the command line,
`$ calrom --config=path/to/my/custom/config`, the standard system-wide and user-specific configuration
files are *not* loaded. Empty config path `$ calrom --config=` makes calrom ignore all configuration
files and use the built-in default configuration.

Example configuration file, loading the proper calendar of the archdiocese of Prague
and disabling colours:

```bash
# shell-like comments can be used in configuration files

--calendar=czech-praha-cs # calendar of the archdiocese of Prague
--calendar=~/calendar_data/local_church.txt # path to a custom calendar file with proper celebrations of the parish where I live (titular feast of the church, dedication)

--load-parents # load also parent calendars specified by the calendar file(s)
               # (default if just one calendar file is specified, but we specified two)

--no-color # disable colours
```

(Configuration file format is inspired by [.rspec][dotrspec], [.yardopts][dotyardopts]
and others.)

Most options work in such a way that if several conflicting options are provided,
the last one wins. You can thus e.g. set your favourite display format (e.g. `--list`)
or date range (e.g. `-3`) in the configuration file and override it, if necessary,
by providing some other option from the same group on the command line.
An exception from this rule is the `--calendar=` option, repeated occurrences of which
do not cancel each other, but are all composed together in the given order to build a calendar
by layering.
Also repeated occurrences of the `--day-filter=` and `--celebration-filter=` options
don't cancel each other, but all specified filtering expressions are applied.

## Running tests

Clone the repository, `$ bundle install` to install dependencies, then:

`$ rake cucumber` - run specs describing the command line interface

`$ rake spec` - run specs describing internal implementation details

`$ rake` - run all groups of specs one after another

## Project roadmap

* [x] detailed listing of a day/month/year/range of dates
* [x] month/year overview - options and output mostly mimicking
  the BSD Unix [`cal`][cal] utility,
  but with liturgical colours and celebration ranks
* [x] condensed format (but with detailed information) suitable for awesome/i3 toolbars etc.
* [x] machine-readable detailed listing
* [ ] year summary: lectionary cycles, movable feasts
* [x] configuration file to set default options
* [x] specify calendar data path (with support for layering several calendars)
* [ ] option to auto-select one of optional celebrations - with multiple supported strategies (prefer ferial, take first non-ferial, configured whitelist, blacklist)
* [ ] integrate online data sources

## Backward compatibility

Project adheres to [semantic versioning][semver]
with regard to the command line interface:
between major releases,
the same configuration (through command line options and other
ways of configuration the application will eventually support)
should print information of the same (or greater) level of detail
for the same range of dates.
For output formats explicitly documented as machine-readable,
format must be preserved (where some machine-readable
formats, especially the structured ones, allow backward-compatible
extensions, others do not, according to their nature).

No backward compatibility is guaranteed on the level of the application's
internal interfaces, since they are not intended to be used
by third-party code.

## CLI patterns

When designing new elements (options, arguments) of the
command line interface

* [(BSD version of) `cal`][cal] should be mimicked where reasonable
* the [Command-Line Options][taoup] chapter from E. S. Raymond's
  *The Art of Unix Programming* should be consulted in the rest
  of cases

## License

GNU/GPL 3.0 or later

[caro]: https://github.com/igneus/calendarium-romanum
[carodata]: https://github.com/igneus/calendarium-romanum/tree/master/data
[calapidoc]: http://calapi.inadiutorium.cz/api-doc
[semver]: https://semver.org/
[cal]: https://www.freebsd.org/cgi/man.cgi?query=cal
[taoup]: http://www.catb.org/esr/writings/taoup/html/ch10s05.html
[dotrspec]: https://relishapp.com/rspec/rspec-core/v/2-0/docs/configuration/read-command-line-configuration-options-from-files
[dotyardopts]: https://rubydoc.info/gems/yard/file/docs/GettingStarted.md#yardopts-options-file


CHANGELOG.md
# Changelog

## [0.4.0] 2022-06-27

### Added

- date range selection: cal-like option `-3` to select the current month together
  with the preceding and following one (contributed by Fabio Soares @fabiosoaresv)
- if exactly one sanctorale data file is specified and its YAML front matter
  references a parent calendar (key `extends`), the file is loaded with its complete
  graph of parents
- option `--[no-]load-parents` to explicitly enable/disable loading of parent calendars
- basic support for remote calendars: the `--calendar=` option now accepts
  also a calendar API URL like
  `--calendar=http://calapi.inadiutorium.cz/api/v0/en/calendars/general-la`,
  the referenced API is expected to be compatible with
  [Liturgical Calendar API](http://calapi.inadiutorium.cz/) v0
- calendar listing (`--calendars`) marks default calendar as such
- calendar listing lists each calendar's parents
- option `--verbose`
- condensed view (`--format=condensed`), intended for use cases like printing current day
  in a window manager toolbar and the like
- option `--to-sunday=` to configure which temporale solemnities should be transferred to Sunday
  (e.g. `--to-sunday=epiphany`)
- option `--temporale-extension=` to activate temporale extension
  (supported values are [temporale extension](https://github.com/igneus/calendarium-romanum/tree/master/lib/calendarium-romanum/temporale/extensions)
  class names, e.g. `--temporale-extension=ChristEternalPriest`)
- any class in the `Calrom::Formatter` module is considered formatter and its
  lowercased name can be used as value for the `--format=` option - custom formatters
  can be added by defining new classes in the namespace
- options `--day-filter=` and `--celebration-filter=` for filtering days and celebrations
  by a Ruby expression
- environment variable `CALROM_CURRENT_DATE` can be used to override current date
  (intended mainly for testing)

### Changed

- automated locale selection based on sanctorale data file relies solely on the `locale` key
  of the file's YAML front matter (previously it relied on a file naming convention)
- option `--calendar=` expands tilde in path (if it isn't expanded by shell,
  as is the case e.g. in configuration files)
- when displaying just a single day, highlighting of the current day is disabled by default
  (i.e. if the day being displayed is current day, it won't be highlighted as such)
- list view (`--list`): rank is not printed for ferials and Sundays
- overview view always arranges the displayed months in multiple columns, just like `cal`
  (previously date ranges spanning less than three months were printed in a single column)
- date range specification: support for dates with year higher than 9999 (e.g. `10000-01-01`)

### Fixed

- crash on invalid value of `--locale`
- list view: when printing a date range between the same months of different years
  (e.g. 2000-01-xx .. 2001-01-xx), year was not printed in month headings

## [0.3.0] 2020-06-21

### Added

- date range selection: options `--today`, `--tomorrow`, `--yesterday`
- specifying the `--calendar=` option multiple times layers the calendars
- file path can be provided as argument to the `--calendar=` option
- option `--locale=` to specify language of localized calendar strings
- language of the (last loaded) sanctorale data is by default used for localized strings
- output is by default not colorized when not printing to a terminal
- configuration files
  - `/etc/calromrc`
  - `$HOME/.calromrc`
  - specified by the `--config=` option
- machine-readable output formats `--format=csv`, `--format=json`
- option `--calendars` to list bundled calendars

### Fixed

- crash on option parsing errors other than invalid option

## [0.2.0] 2020-06-16

### Added

- cal-like view (is now default for month, year and date range listing)
- support more cal-like ways of specifying date range
  - month argument `calrom MONTH YEAR` (as an alternative to month option `cal -m MONTH YEAR`)
  - month option with p/f flag `calrom -m 5p` / `calrom -m 5f` as a shortcut to specify a month of the previous/following year
  - option to print the whole current year `calrom -y`
- support arbitrary date range specified as two ISO dates
- input is validated, errors handled and reported
- cal-like option to print (only) the date of Easter
- cal-like debugging options `-d` and `-H`
- option `--[no-]color` to enable/disable ANSI colours (enabled by default)

### Changed

- list view: contains weekday names; month names instead of month numbers

### Fixed

- unhandled edge-case in `Calrom::Month` when dealing with the end of a year
- liturgical colour violet

## [0.1.0] 2019-10-05

### Added

- lists (general Roman) calendar (in English) for the current month
- liturgical colours, celebration rank highlighting
- highlighting of the current day
- date range specification: `cal`-like year and month; single ISO date


lib/calrom/option_parser.rb
require 'optparse'

module Calrom
  # Parses command line options, produces Config.
  class OptionParser
    # Raised when the options being parsed are invalid.
    #
    # Exceptions raised by the option parsing library are translated to this exception,
    # thus hiding implementation details (the option parsing library's exception types)
    # from the outer world.
    class Error < RuntimeError; end

    class CustomizedOptionParser < ::OptionParser
      def separator(string)
        super "\n" + string
      end
    end

    def self.call(*args)
      self.new.call(*args)
    end

    def call(argv, config = nil)
      config ||= Config.new

      today = config.today

      year = today.year
      month = today.month
      day = nil
      another_day = nil

      range_type = nil

      opt_parser = CustomizedOptionParser.new do |opts|
        opts.banner = <<~EOS
        Usage: calrom [options] [arg1 [arg2]]

        Specifying date range (cal/ncal-compatible):

          calrom             - current month
          calrom -m 5        - May of the current year
          calrom -m 5p       - May of the previous year
          calrom -m 5f       - May of the following year
          calrom -m 5 2000   - May 2000
          calrom 5 2000      - also May 2000
          calrom 2000        - whole year 2000
          calrom -y 2000     - also whole year 2000
          calrom -y          - whole current year
          calrom -3          - display the previous, current and next month surrounding today

        Specifying date range (not cal-compatible):

          calrom 2000-05-31              - specified day (only)
          calrom 2000-05-31 2000-07-01   - arbitrary date range
          calrom (--yesterday|--today|--tomorrow)

        EOS

        opts.separator 'Configuration files'

        opts.on('--config=CONFIG', 'load configuration from file (may be used multiple times, all specified files will be loaded)') do |value|
          config.configs << value
        end

        opts.separator 'Options selecting date range'

        # cal
        opts.on('-m MONTH', '--month=MONTH', 'display the specified month. \'f\' or \'p\' can be appended to display the same month of the following or previous year respectively') do |value|
          range_type = :month
          if value =~ /^(\d+)([pf])$/
            month = $1
            year = validate_year(year) + ($2 == 'f' ? 1 : -1)
          else
            month = value
          end
        end

        # cal
        opts.on('-3', 'display the previous, current and next month surrounding today') do |value|
          range_type = :three_months
        end

        # cal
        opts.on('-y', '--year', 'display specified (or current) year') do |value|
          range_type = :year
        end

        opts.on('--yesterday', 'display previous day') do |value|
          day = today - 1
          range_type = :day
        end

        opts.on('--today', 'display current day') do |value|
          day = today
          range_type = :day
        end

        opts.on('--tomorrow', 'display following day') do |value|
          day = today + 1
          range_type = :day
        end

        opts.separator "Options configuring liturgical calendar"

        opts.on('-c CAL', '--calendar=CAL', 'specify (sanctorale) calendar to use. If repeated, layers all specified calendars one over another') do |value|
          config.sanctorale << value
        end

        opts.on('--[no-]load-parents', 'explicitly enable/disable parent calendar loading') do |value|
          config.load_parents = value
        end

        transferable = CR::Temporale::SUNDAY_TRANSFERABLE_SOLEMNITIES
        opts.on('--to-sunday=SOLEMNITY', transferable, 'transfer solemnity to Sunday' + supported_values(transferable)) do |value|
          config.transfer_to_sunday << value.to_sym
        end

        extensions = CR::Temporale::Extensions.all.collect {|cls| cls.name.split('::').last.to_sym }
        opts.on('--temporale-extension=EXTENSION', extensions, 'use temporale extension' + supported_values(extensions)) do |value|
          config.temporale_extensions << CR::Temporale::Extensions.const_get(value)
        end

        opts.on('--locale=LOCALE', 'override language in which temporale celebration titles are rendered' + supported_values(I18n.available_locales)) do |value|
          config.locale = value.to_sym
        end

        opts.separator 'Options affecting presentation'

        opts.on('-l', '--list', 'display detailed listing of days and celebrations (synonym to --format=list)') do
          config.formatter = :list
        end

        supported_formats =
          Calrom::Formatter.constants
            .collect(&:downcase)
            .delete_if {|i| i == :formatter }
        formats_help = supported_formats.join(', ')
        opts.on('--format=FORMAT', supported_formats, "specify output format (supported: #{formats_help})") do |value|
          config.formatter = value
        end

        opts.on('--day-filter=EXPR', 'display only days for which the expression (Ruby snippet executed in context of each CalendariumRomanum::Day instance) evaluates as truthy') do |expr|
          config.filter_days << expr
        end

        opts.on('--celebration-filter=EXPR', 'display only celebrations for which the expression (Ruby snippet executed in context of each CalendariumRomanum::Celebration instance) evaluates as truthy') do |expr|
          config.filter_celebrations << expr
        end

        # cal
        opts.on('-e', '--easter', 'display date of Easter (only)') do
          config.formatter = :easter
        end

        opts.on('--calendars', 'list bundled calendars') do |value|
          config.formatter = :calendars
        end

        opts.on('--[no-]color', 'enable/disable colours (enabled by default)') do |value|
          config.colours = value
        end

        opts.on('-v', '--verbose', 'enable verbose output') do
          config.verbose = true
        end

        opts.separator 'Debugging options'

        # cal
        opts.on('-d YM', '--current-month=YM', 'use given month (YYYY-MM) as the current month (for debugging of date range selection)') do |value|
          year, month = value.split '-'
        end

        # cal
        opts.on('-H DATE', '--highlight-date=DATE', 'use given date as the current date (for debugging of highlighting)') do |value|
          config.today = validate_day value
        end

        opts.separator 'Information regarding calrom'

        opts.on('-V', '--version', 'display calrom version') do
          puts 'calrom v' + Calrom::VERSION
          puts 'using calendarium-romanum v' + CR::VERSION
          exit
        end

        # Normally optparse defines this option by default, but once -H option is added,
        # for some reason -h (if not defined explicitly) is treated as -H.
        opts.on('-h', '--help', 'display this help') do
          puts opts
          exit
        end
      end

      begin
        arguments = opt_parser.parse argv
      rescue ::OptionParser::ParseError => e
        raise Error.new(e.message)
      end

      iso_date_regexp = /^(\d{4,}-\d{2}-\d{2})$/

      case arguments
      in [^iso_date_regexp => date]
        range_type = :day
        day = date
      in [^iso_date_regexp => date, ^iso_date_regexp => another_date]
        range_type = :free
        day = date
        another_day = another_date
      in [y]
        range_type ||= :year
        year = y
      in [m, y]
        range_type = :month
        month, year = m, y
      in []
      else
        raise InputError.new('too many arguments')
      end

      config.date_range =
        build_date_range(
          range_type,
          validate_year(year),
          validate_month(month),
          day && validate_day(day),
          another_day && validate_day(another_day)
        )

      config.freeze
    end

    protected

    def validate_year(year)
      unless year.is_a?(Integer) || year =~ /^\d+$/
        raise InputError.new("not a valid year #{year}")
      end

      year.to_i
    end

    def validate_month(month)
      unless month.is_a?(Integer) ||
             (month =~ /^\d+$/ && (1..12).include?(month.to_i))
        raise InputError.new("not a valid month #{month}")
      end

      month.to_i
    end

    def validate_day(day)
      return day if day.is_a? Date

      Date.parse(day)
    rescue ArgumentError
      raise InputError.new("not a valid date #{day}")
    end

    def build_date_range(range_type, year, month, day, another_day=nil)
      range =
        case range_type
        when :year
          Year.new(year)
        when :day
          Day.new(day)
        when :free
          DateRange.new(day, another_day)
        when :three_months
          ThreeMonths.new(year, month)
        else
          Month.new(year, month)
        end

      beginning = CR::Calendar::EFFECTIVE_FROM
      if range.first < beginning
        raise InputError.new("implemented calendar system in use only since #{beginning}")
      end

      range
    end

    def supported_values(values)
       " (supported: #{values.join(', ')})"
    end
  end
end
